### PR TITLE
New utilities & move things around

### PR DESCRIPTION
**Pull request by ikitommi**
* `malli.util/equals` (no-doc)
* `malli.util/simplify-map-entry` (no-doc)
* `malli.util/merge`
* `malli.util/union`

* `malli.core/required-map-entry?` (no-doc)
* `malli.core/map-entries` (`:map` schemas implement new `malli.core/MapSchema` protocol)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -404,31 +404,61 @@ Single sweep of defaults & string encoding:
...
+     [:street string?]
+     [:country [:enum "finland" "poland"]]]]]
+  [:map
+   [:description {:optional true} string?]
+   [:address
+    [:map
+     [:country string?]]]])
+;[:map
+; [:name string?]
+; [:description string?]
```
Comment (rschmukler) [inline README.md]: Should this be optional?

**Review by rschmukler — CHANGES_REQUESTED** [review]
Edit: Looks like I misread the code - Looks good!

I do have a question as to whether `union` should cause optional to be transferred to the new schema. I would expect it would since any value of `A` or `B`, including the map missing the field should be valid. The issue is then that incorrect values of `B` could be accepted. I'm wondering which behavior is more surprising. Maybe the only "safe" answer is using `[:or schema-1 schema-2]`?

File: src/malli/core.cljc [hunk 2]
```diff
@@ -185,36 +175,33 @@
...
-     :forms (mapv (fn [[k p v]]
-                    (let [v' (-form v)]
-                      (if p [k p v'] [k v']))) entries)}))
+(defn- -parse-map-entries [children opts]
+  (->> children (keep identity) (mapv #(-expand-key % opts identity))))
+
+(defn ^:no-doc map-entry-forms [entries]
+  (mapv (fn [[k p v]] (let [v' (-form v)] (if p [k p v'] [k v']))) entries))
+
+(defn ^:no-doc required-map-entry [[_ ?p]]
```
Comment (rschmukler) [inline src/malli/core.cljc]: Consider renaming `required-map-entry?` so it's obvious that it's a predicate. Also perhaps this should not be `no-doc`?

File: README.md (hunk 1 above)
Comment (ikitommi) [reply README.md]: I think it should, fixed.

File: src/malli/core.cljc (hunk 2 above)
Comment (ikitommi) [reply src/malli/core.cljc]: renamed but left the `:no-doc` for now.

**Comment by ikitommi** [comment]
moved `malli.core/equals` into a no-doc function in `malli.util` as it's only used in there. Might need to rethink Schema equality (equals & hash-code?), also less public api for now.

**Comment by ikitommi** [comment]
btw, `util` is a bad name for namespace, but then again `mu` was not used as an ns alias...

``` clj
(require '[malli.util :as mu])
```

**Comment by ikitommi** [comment]
thanks for the comments @rschmukler !

**Pull request merged**